qgis/examples/elevation_tile_downloader/gui/tile_downloader_dlg.py

```diff
@@ -0,0 +1,148 @@
+"""
+from pathlib import Path
+
+from qgis.PyQt.QtCore import Qt
+from qgis.PyQt.QtWidgets import QDialogButtonBox, QProgressBar, QDialog
+from qgis.gui import QgisInterface, QgsMessageBarItem
+from qgis.core import (
+    QgsProject,
+    QgsVectorLayer,
+    QgsGeometry,
+    QgsMapLayerProxyModel,
+    QgsApplication,
+    QgsRectangle,
+    QgsCoordinateTransform,
+    QgsCoordinateReferenceSystem,
+    Qgis,
+    QgsTask,
+)
+from math import floor, ceil
+from typing import List, Optional
+
+from ..gui.tile_downloader_dlg_base import Ui_ElevationTileDownloaderDialogBase
+from ..core.task import DownloadTask
+
+
+def to_wgs84(
+    geometry: QgsGeometry, own_crs: QgsCoordinateReferenceSystem
+) -> QgsGeometry:
+    """
+    Transforms the ``point`` to (``direction=ForwardTransform``) or from
+    (``direction=ReverseTransform``) WGS84.
+    """
+    wgs84 = QgsCoordinateReferenceSystem.fromEpsgId(4326)
+    if own_crs != wgs84:
+        xform = QgsCoordinateTransform(own_crs, wgs84, QgsProject.instance())
+        geometry.transform(xform)
+    return geometry
+
+
+def create_grid_from_bounds(bounds: QgsRectangle) -> List[QgsGeometry]:
+    """Creates a regular grid of size 1x1 within specified bounds"""
+    grid_bboxes = []
+
+    # loop through x and y range and create the grid
+    min_x, min_y = [floor(x) for x in (bounds.xMinimum(), bounds.yMinimum())]
+    max_x, max_y = [
+        ceil(x)
+        for x in (
+            bounds.xMinimum() + bounds.width(),
+            bounds.yMinimum() + bounds.height(),
+        )
+    ]
+    for x in range(min_x, max_x):
+        for y in range(min_y, max_y):
+            grid_bboxes.append(
+                QgsGeometry.fromRect(QgsRectangle(x, y, x + 1, y + 1))
+            )
+
+    return grid_bboxes
+
+
+class ElevationTileDownloaderDialog(
+    QDialog, Ui_ElevationTileDownloaderDialogBase
+):
+    def __init__(self, iface: QgisInterface, parent=None):
+        """Constructor."""
+        super(ElevationTileDownloaderDialog, self).__init__(parent)
+        self.iface = iface
+        self.setupUi(self)
+
+        self.progress_bar: Optional[QProgressBar] = None
+        self.task: Optional[QgsTask] = None
+        self.total = 0
+
+        # we only allow polygon layers
+        self.layer_choice.setFilters(QgsMapLayerProxyModel.PolygonLayer)
+        self.accepted.connect(self.download_tiles)
+
+        # change the text of the "Ok" button
+        self.button_box.button(QDialogButtonBox.Ok).setText("Download")
+
+    def download_tiles(self) -> None:
+        """
+        The plugin's main method; constructs the grid from selected input features,
+        creates and starts the download task and connects the plugin to task signals.
+        """
+        poly_layer: QgsVectorLayer = self.layer_choice.currentLayer()
+        out_dir = Path(self.output_dir.filePath())
+
+        # Collect the tile bounding boxes covering the input polygons
+        grid = set()
```

Comment:
if not `set` we could get duplicates if overlapping polygons intersect the grid I think